### PR TITLE
fix(gpu): use all gpus in omp loops

### PR DESCRIPTION
**Pull request by agnesLeroy**
closes: https://github.com/zama-ai/tfhe-rs-internal/issues/606

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: backends/tfhe-cuda-backend/cuda/include/integer.h [hunk 1]
```diff
@@ -1535,8 +1535,8 @@ template <typename Torus> struct int_arithmetic_scalar_shift_buffer {
 
   Torus *tmp_rotated;
 
-  cudaStream_t local_stream_1;
-  cudaStream_t local_stream_2;
+  cudaStream_t *local_streams_1;
+  cudaStream_t *local_streams_2;
```
Comment (pdroalves) [inline backends/tfhe-cuda-backend/cuda/include/integer.h:1539]: Shouldn't we use vectors for consistency with `int_radix_lut`?

File: backends/tfhe-cuda-backend/cuda/include/integer.h [hunk 2]
```diff
@@ -2119,8 +2133,8 @@ template <typename Torus> struct int_comparison_buffer {
 
   // Used for scalar comparisons
   int_radix_lut<Torus> *signed_msb_lut;
-  cudaStream_t lsb_stream;
-  cudaStream_t msb_stream;
+  cudaStream_t *lsb_streams;
+  cudaStream_t *msb_streams;
```
Comment (pdroalves) [inline backends/tfhe-cuda-backend/cuda/include/integer.h:2137]: Same

File: backends/tfhe-cuda-backend/cuda/include/integer.h [hunk 3]
```diff
@@ -2316,10 +2338,10 @@ template <typename Torus> struct int_div_rem_memory {
...
 
   // sub streams
-  cudaStream_t sub_stream_1;
-  cudaStream_t sub_stream_2;
-  cudaStream_t sub_stream_3;
-  cudaStream_t sub_stream_4;
+  cudaStream_t *sub_streams_1;
+  cudaStream_t *sub_streams_2;
+  cudaStream_t *sub_streams_3;
+  cudaStream_t *sub_streams_4;
```
Comment (pdroalves) [inline backends/tfhe-cuda-backend/cuda/include/integer.h:2344]: Same

File: backends/tfhe-cuda-backend/cuda/src/integer/cmux.cuh [hunk 4]
```diff
@@ -51,31 +51,35 @@ __host__ void host_integer_radix_cmux_kb(
 
   // Since our CPU threads will be working on different streams we shall assert
   // the work in the main stream is completed
-  cuda_synchronize_stream(streams[0], gpu_indexes[0]);
-  auto true_stream = mem_ptr->zero_if_true_buffer->local_stream;
-  auto false_stream = mem_ptr->zero_if_false_buffer->local_stream;
+  auto true_streams = mem_ptr->zero_if_true_buffer->local_streams;
+  auto false_streams = mem_ptr->zero_if_false_buffer->local_streams;
+  for (uint j = 0; j < gpu_count; j++) {
+    cuda_synchronize_stream(streams[j], gpu_indexes[j]);
```
Comment (pdroalves) [inline backends/tfhe-cuda-backend/cuda/src/integer/cmux.cuh:57]: Why do we need to synchronize any device other than 0?

**Review by pdroalves — APPROVED** [review]
Nice PR. I just added a design question. We can merge as it is if you wish.

File: backends/tfhe-cuda-backend/cuda/include/integer.h (hunk 1 above)
Comment (agnesLeroy) [reply backends/tfhe-cuda-backend/cuda/include/integer.h:1539]: Yes, I'll change it :+1:

Comment (agnesLeroy) [reply backends/tfhe-cuda-backend/cuda/include/integer.h:1539]: Ok so it's not so easy to change: we pass streams with `cudaStream_t *` type that cross the c++/Rust interface, so it's much easier to keep it this way.

File: backends/tfhe-cuda-backend/cuda/src/integer/cmux.cuh (hunk 4 above)
Comment (agnesLeroy) [reply backends/tfhe-cuda-backend/cuda/src/integer/cmux.cuh:57]: We're going to be executing on different streams on each GPU so I bellieve we have to first synchronize the "main" stream on each GPU as well. It's the same logic we had before, except across multiple GPUs.

**Pull request merged**